+++ a/common/hooks.c	2017-09-23 05:30:38.617933828 +0200
+++ b/common/hooks.c	2017-09-23 05:30:02.431314299 +0200
@@ -64,6 +64,7 @@
 #include <sys/syscall.h>
 #include <sys/auxv.h>
 #include <sys/prctl.h>
+#include <sys/uio.h>
 
 #include <sys/mman.h>
 #include <libgen.h>
@@ -2624,7 +2625,7 @@
     HOOK_DIRECT(malloc),
     HOOK_DIRECT_NO_DEBUG(free),
     HOOK_DIRECT_NO_DEBUG(calloc),
-    HOOK_DIRECT_NO_DEBUG(cfree),
+    HOOK_DIRECT_NO_DEBUG(free),
     HOOK_DIRECT_NO_DEBUG(realloc),
     HOOK_DIRECT_NO_DEBUG(memalign),
     HOOK_DIRECT_NO_DEBUG(valloc),
