From b3844eae4e93c73e82bb8d40ca666856f3ecb3d4 Mon Sep 17 00:00:00 2001
From: Rainer Hochecker <fernetmenta@online.de>
Date: Fri, 23 Jun 2017 08:54:30 +0200
Subject: [PATCH] VideoPlayer: flush renderer on channel switch

---
 xbmc/cores/VideoPlayer/VideoPlayer.cpp | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/xbmc/cores/VideoPlayer/VideoPlayer.cpp b/xbmc/cores/VideoPlayer/VideoPlayer.cpp
index 4dff63b461b6..26058fd60032 100644
--- a/xbmc/cores/VideoPlayer/VideoPlayer.cpp
+++ b/xbmc/cores/VideoPlayer/VideoPlayer.cpp
@@ -2864,6 +2864,7 @@ void CVideoPlayer::HandleMessages()
              m_messenger.GetPacketCount(CDVDMsg::PLAYER_CHANNEL_SELECT_NUMBER) == 0)
     {
       FlushBuffers(DVD_NOPTS_VALUE, true, true);
+      m_renderManager.Flush();
       CDVDInputStreamPVRManager* input = dynamic_cast<CDVDInputStreamPVRManager*>(m_pInputStream);
       //! @todo find a better solution for the "otherStreamHack"
       //! a stream is not supposed to be terminated before demuxer
@@ -2892,6 +2893,7 @@ void CVideoPlayer::HandleMessages()
              m_messenger.GetPacketCount(CDVDMsg::PLAYER_CHANNEL_SELECT) == 0)
     {
       FlushBuffers(DVD_NOPTS_VALUE, true, true);
+      m_renderManager.Flush();
       CDVDInputStreamPVRManager* input = dynamic_cast<CDVDInputStreamPVRManager*>(m_pInputStream);
       if (input && input->IsOtherStreamHack())
       {
@@ -2926,6 +2928,7 @@ void CVideoPlayer::HandleMessages()
         {
           g_infoManager.SetDisplayAfterSeek(100000);
           FlushBuffers(DVD_NOPTS_VALUE, true, true);
+          m_renderManager.Flush();
           if (input->IsOtherStreamHack())
           {
             CloseDemuxer();
